physlr/physlr.py

```diff
@@ -1021,107 +1020,127 @@ def assign_read_molecule(mxs, moltomxs, mol_counts, bx):
+            barcode_dict = dict(zip(range(len(node_set)), list(node_set)))
+            edges_to_remove_barcode = [(barcode_dict[i], barcode_dict[j])
+                                       for i, j in edges_to_remove]
+            sub_graph_copy = nx.Graph(g.subgraph(node_set))
+            sub_graph_copy.remove_edges_from(edges_to_remove_barcode)
+            cos_components = list(nx.connected_components(sub_graph_copy))
+            cos_components.sort(key=len, reverse=True)
+            for com in cos_components:
+                # if len(com) > 1:
+                communities.append(com)
+        return communities
+
+    @staticmethod
+    def determine_molecules_bc_k_cliques(g, u):
+        """
+        Assign the neighbours of this vertex to molecules by
+        applying k-cliques community detection for each bi-connected component.
+        """
+        return [community
+                for bi_connected_component in
+                Physlr.community_detection_biconnected_components(g, set(g.neighbors(u)))
+                for community in
+                Physlr.community_detection_k_clique(g, bi_connected_component)]
+
+    @staticmethod
+    def determine_molecules_bc_louvain(g, u):
+        """
+        Assign the neighbours of this vertex to molecules by
+        applying louvain for each bi-connected component.
+        """
+        return [community
+                for bi_connected_component in
+                Physlr.community_detection_biconnected_components(g, set(g.neighbors(u)))
+                for community in
+                Physlr.community_detection_louvain(g, bi_connected_component)]
+
+    @staticmethod
+    def determine_molecules_bc_cosine_of_squared(g, u):
+        """
+        Assign the neighbours of this vertex to molecules by
+        applying cosine of squared of the adjacency matrix for each bi-connected component.
+        """
+        return [community
+                for bi_connected_component in
+                Physlr.community_detection_biconnected_components(g, set(g.neighbors(u)))
+                for community in
+                Physlr.community_detection_cosine_of_squared(g, bi_connected_component)]
 
     @staticmethod
     def determine_molecules(g, u, strategy):
-        "Assign the neighbours of this vertex to molecules."
-        if strategy == 2:
-            return Physlr.determine_molecules_k_clique_communities(g, u)
-        if strategy == 3:
-            return Physlr.determine_molecules_louvain(g, u)
-        if strategy == 4:
-            return Physlr.determine_molecules_cosine_of_squared(g, u)
+        """Assign the neighbours of this vertex to molecules."""
+        communities = []
+        if strategy not in {2, 3, 4, 5, 10}:
+            # default strategy (bi-connected components) in case of wrong parameter.
+            communities = \
+                Physlr.community_detection_biconnected_components(g, set(g.neighbors(u)))
```

Comment:
This should produce an error message if the parameter value is not valid.

Reply:
aha, and what function should I use to halt/exit the program @sjackman ? Should I call it inside this function?

Reply:
Yes. `exit(1)`